File: Iteration-2/src/MandatoryCourse.java
import java.util.ArrayList;
import org.apache.log4j.Logger;

public class MandatoryCourse extends Course {
    private Logger log = Logger.getLogger(MandatoryCourse.class);
    private int semester;
    private ArrayList<Course> prerequisites;

    public MandatoryCourse(String courseName, String courseCode, int courseCredit, int courseDay, String courseHour,
            int quota, int semester, ArrayList<Course> prerequisites) {
        super(courseName, courseCode, courseCredit, courseDay, courseHour, quota);
        this.semester = semester;
        this.prerequisites = prerequisites;
        log.info(this.getCourseName() + " (" + this.getCourseCode() + ")"
                + " named mandatory course created.");
    }

    @Override
    public boolean isEligibleToRequest(Student student) {
        if (student.getSemester() == this.getSemester()) {
            if (!student.getTranscript().hasBeenPassedCourses(this.getPrequisites())) {
                this.setFailedPreq(this.getFailedPreq() + 1);
                student.getStudentOutput().add("The system didn't allow " + this.getCourseCode() +
                        " because student failed prereq. " + this.getPrequisites().get(0).getCourseCode());
                log.info(
                        "Student couldn't pass prerequisite course(s). " + student.getStudentName() + " cannot enroll "
                                + this.getCourseName()
                                + ".");
                return false;
            } else {
                /* log.info(
                        "Student passed prerequisite course(s) successfully. " + student.getStudentName()
                                + " can enroll "
                                + this.getCourseName()
                                + "."); */
                return true;
            }
        } else {
            log.info(
                    "Student is not eligible to take this course this semester. " + student.getStudentName()
                            + " cannot enroll "
                            + this.getCourseName()
                            + ".");
            return false;
        }
    }

    public boolean isEligibleToBePreviouslyTaken(Student student) {
        /* log.info(student.getStudentName() + " could "
                + ((student.getSemester() > this.getSemester()
                        && student.getTranscript().hasBeenPassedCourses(this.getPrequisites()))
                                ? ""
                                : "not")
                + "take this course for his previous semesters: "
                + this.getCourseName()); */
        return student.getSemester() > this.getSemester()
                && student.getTranscript().hasBeenPassedCourses(this.getPrequisites());
    }

    // GETTER & SETTER
    public int getSemester() {
        return semester;
    }

    public ArrayList<Course> getPrequisites() {
        return prerequisites;
    }
}

File: Iteration-3/Code/MandatoryCourse.py
from Course import Course
from typing import List


class MandatoryCourse(Course):
    def __init__(self, courseName: str, courseCode: str, courseCredit: int, courseDay: int, courseHour: str,
                 quota: int, semester: int, prerequisites: List[Course]):
        super().__init__(courseName, courseCode, courseCredit, courseDay, courseHour, quota)
        self.__semester = semester
        self.__prerequisites = prerequisites

    def isEligibleToRequest(self, student):
        if (student.getSemester() == self.getSemester()):
            if (not student.getTranscript().hasBeenPassedCourses(self.getPrerequisites())):
                self.setFailedPreq(self.getFailedPreq() + 1)
                # student.getStudentOutput().append("The system didn't allow " + self.getCourseCode() + "because student failed prereq." + self.getPrequisites()[0].getCourseCode())
                student.getStudentOutput().append("The system didn't allow " + self.getCourseCode() + " because student failed prereq.")
                return False
            else:
                return True
        else:
            return False

    def isEligibleToBePreviouslyTaken(self, student):
        if student.getSemester() > self.getSemester():
            if student.getTranscript().hasBeenPassedCourses(self.getPrerequisites()):
                return True
            else:
                return False
        else:
            return False
        # return (student.getSemester() > self.getSemester()) and (student.getTranscript().hasBeenPassedCourses(self.getPrerequisites()))

    # GETTERS AND SETTERS
    def getSemester(self) -> int:
        return self.__semester

    def getPrerequisites(self) -> List[Course]:
        return self.__prerequisites